<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''select 
                        count(*) vacancies_count
                from 
                        public.VACANCIES 
            '''

In [5]:
# результат запроса
pd.read_sql_query(query_3_1, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1469115235.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_1, connection)


vacancies_count
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select 
                        count(*) employers_count
                from 
                        public.EMPLOYERS 
            '''

In [7]:
# результат запроса
pd.read_sql_query(query_3_2, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/4057088587.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_2, connection)


employers_count
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select 
                        count(*) areas_count
                from 
                        public.AREAS 
            '''

In [9]:
# результат запроса
pd.read_sql_query(query_3_3, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/4147122770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_3, connection)


areas_count
0         1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select 
                        count(*) industries_count 
                from 
                        public.INDUSTRIES 
            '''

In [11]:
# результат запроса
pd.read_sql_query(query_3_4, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2627729322.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_4, connection)


industries_count
0               294

***

#### Выводы по предварительному анализу данных
Количество вакансий превышает количество работодателей приблизительно в 2 раза.
В среднем на одну сферу приходится по 80 работодателей.
Также в среднем на один регион приходится около 36 вакансий.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# левая таблица с вакансиями, т.к. в рамках данной задачи не рассматриваем регионы без вакансий
query_4_1 = f'''select 
                        a.name area,
                        count(v.area_id) cnt
                from 
                        public.vacancies v
                        left join public.areas a on v.area_id = a.id
                group by 
                        a.name
                order by 
                        count(v.area_id) desc
                limit (5)
            '''

In [13]:
# результат запроса
pd.read_sql_query(query_4_1, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/4222419522.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''select 
                        count(v.id) vac_salary_not_empty
                from 
                        public.vacancies v
                where
                        v.salary_from is not null or v.salary_to is not null
            '''

In [15]:
# результат запроса
pd.read_sql_query(query_4_2, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/614600247.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_2, connection)


vac_salary_not_empty
0                 24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query_4_3 = f'''select 
                        round(avg(v.salary_from),0) sal_avg_from,
                        round(avg(v.salary_to),0) sal_avg_to
                from 
                        public.vacancies v
            '''
            

In [17]:
# результат запроса
pd.read_sql_query(query_4_3, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3534961570.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_3, connection)


sal_avg_from  sal_avg_to
0       71065.0    110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''select 
                        v.schedule,
                        v.employment,
                        count(v.id) vacancies_count
                from 
                        public.vacancies v
                group by 
                        v.schedule,
                        v.employment
                order by 
                        count(v.id) desc
            '''

In [19]:
# результат запроса
pd.read_sql_query(query_4_4, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2630570843.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancies_count
0        Полный день     Полная занятость            35367
1   Удаленная работа     Полная занятость             7802
2      Гибкий график     Полная занятость             1593
3   Удаленная работа  Частичная занятость             1312
4     Сменный график     Полная занятость              940
5        Полный день           Стажировка              569
6     Вахтовый метод     Полная занятость              367
7        Полный день  Частичная занятость              347
8      Гибкий график  Частичная занятость              312
9        Полный день     Проектная работа              141
10  Удаленная работа     Проектная работа              133
11     Гибкий график           Стажировка              116
12    Сменный график  Частичная занятость              101
13  Удаленная работа           Стажировка               64
14     Гибкий график     Проектная работа               18
15    Сменный график           Стажировка               12
16    Вахтовый метод     Проектная работа                2
17    Сменный график     Проектная работа                1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query_4_5 = f'''select 
                        distinct v.experience,
                        count(v.id) vac_count
                from 
                        public.vacancies v
                group by
                        1
                order by
                        2
            '''

In [21]:
# результат запроса
pd.read_sql_query(query_4_5, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3190483664.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_5, connection)


experience  vac_count
0         Более 6 лет       1337
1           Нет опыта       7197
2       От 3 до 6 лет      14511
3  От 1 года до 3 лет      26152

***

#### выводы по детальному анализу вакансий
Практически четверть всех вакансий сосредоточена всего в 5 наиболее крупных городах.
Ориентир по зарплате указан лишь в половине вакансий.
В среднем заявленные зарплаты находятся в диапазоне 71-110т.р.
Наиболее популярный предлагаемый режим работы - полная занятость в удаленном формате.
При этом, подавляющее большинство работодателей ищут сотрудников либо с опытом от года до трех лет, либо от 3 лет до 6.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
query_5_1 = f'''select 
                        distinct e.name employer,
                        count(v.id) vac_count
                from 
                        public.vacancies v
                        left join public.employers e on v.employer_id=e.id
                group by
                        1
                order by
                        2 desc
                limit (5)
            '''

In [23]:
# результат запроса
pd.read_sql_query(query_5_1, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3902369145.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1, connection)


employer  vac_count
0         Яндекс       1933
1     Ростелеком        491
2       Тинькофф        444
3           СБЕР        428
4  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# основная таблица с регионами, к которой левым соединением добавляем работодателей и отфильтрованные вакансии
query_5_2 = f'''select 
                        distinct a.name area,
                        count(distinct e.id) empl_count,
                        count(v.id) vac_count
                from 
                        public.areas a
                        left join public.employers e on e.area=a.id
                        left join public.vacancies v on v.area_id=a.id
                where
                        v.id is null
                group by
                        1
                order by
                        2 desc
                limit(1)
            '''

In [25]:
# результат запроса
pd.read_sql_query(query_5_2, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2146817738.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_2, connection)


area  empl_count  vac_count
0  Россия         410          0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# основная таблица Работодатели, к которой левым соединением добавляем таблицу с вакансиями
query_5_3 = f'''select 
                        e.name employer,
                        count(distinct v.area_id) area_count
                from 
                        public.employers e
                        left join public.vacancies v on e.id=v.employer_id
                group by
                        e.id
                order by
                        2 desc
            '''

In [27]:
# результат запроса
pd.read_sql_query(query_5_3, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/593965358.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_3, connection)


employer  area_count
0                      Яндекс         181
1                  Ростелеком         152
2                  Спецремонт         116
3      Поляков Денис Иванович          88
4                   ООО ЕФИН           71
...                       ...         ...
23496              Но Смокинг           0
23497        МАСТЕРСКАЯ CREED           0
23498           Dzing Finance           0
23499                    КОРС           0
23500             Наука-Связь           0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# основная таблица Работодатели, к которой левым соединением добавляем таблицу со сферами для них
query_5_4 = f'''select 
                        count(e.id) empl_wo_ind_count
                from 
                        public.employers e
                        left join public.EMPLOYERS_INDUSTRIES e_i on e.id=e_i.employer_id
                where
                        e_i.industry_id is null
            '''

In [29]:
# результат запроса
pd.read_sql_query(query_5_4, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1644714913.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_4, connection)


empl_wo_ind_count
0               8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# основная таблица СферыРаботодателей, к которой левым соединением добавляем таблицу работодателей (с названиями)
query_5_5 = f'''select 
                        e.name employer,
                        count(e_i.industry_id) empl_4_ind_count
                from 
                        public.EMPLOYERS_INDUSTRIES e_i
                        left join public.employers e on e.id=e_i.employer_id
                group by
                        e.name
                having
                        count(e_i.industry_id)=4
                order by
                        1
                offset(2)
                limit(1)
            '''

In [31]:
# результат запроса
pd.read_sql_query(query_5_5, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1169902027.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_5, connection)


employer  empl_4_ind_count
0     2ГИС                 4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# основная таблица СферыРаботодателей, к которой левым соединением добавляем таблицу со Сферами (для фильтра по названию)
query_5_6 = f'''select 
                        count(e_i.employer_id) software_dev_count
                from 
                        public.EMPLOYERS_INDUSTRIES e_i
                        left join public.industries i on e_i.industry_id=i.id
                where
                        i.name='Разработка программного обеспечения'
            '''

In [33]:
# результат запроса
pd.read_sql_query(query_5_6, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2878008418.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_6, connection)


software_dev_count
0                3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
# модуль PYTHON-17 бонусный, изучение отложено на свободное время, когда буду идти с опережением графика
# предполагаю, что отсутствие данного пункта не должно влиять на итоговую оценку

mln = (
    'Москва',
    'Санкт-Петербург',
    'Новосибирск',
    'Екатеринбург',
    'Казань',
    'Нижний Новгород',
    'Красноярск',
    'Челябинск',
    'Самара',
    'Уфа',
    'Ростов-на-Дону',
    'Омск',
    'Краснодар',
    'Воронеж',
    'Пермь',
    'Волгоград'
)


In [35]:
query_5_7 = f'''(select 
                        a.name region,
                        count(v.id) vacancies_count
                from 
                        public.vacancies v
                        left join public.employers e on e.id=v.employer_id
                        left join public.areas a on a.id=v.area_id
                where
                        e.name='Яндекс' and a.name in {mln}
                group by
                        a.name
                order by
                        2 desc)
                union
                (select 
                        'total' as region,
                        count(v.id) as vacancies_count
                from 
                        public.vacancies v
                        left join public.employers e on e.id=v.employer_id
                        left join public.areas a on a.id=v.area_id
                where
                        e.name='Яндекс' and a.name in {mln})
                order by vacancies_count
            '''

In [36]:
# результат запроса
pd.read_sql_query(query_5_7, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1216511710.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_7, connection)


region  vacancies_count
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5            Казань               25
6    Ростов-на-Дону               25
7               Уфа               26
8            Самара               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            total              485

***

#### выводы по анализу работодателей
Среди всех работодателей сильно выделяется Яндекс, у которого в несколько раз больше вакансий, чем у любого другого работодателя.
Имеется 1148 работодателей в различных регионах, которые вообще не разместили вакансии.
В выборке присутствуют компании, которые размещают вакансии не только в регионе своего офлайн-присутствия,
вплоть до 181 вакансии в разных регионах.
Далеко не все работодатели указывают свою сферу деятельности (8,4тыс.), что может мешать кандидатам качественно оценивать вакансии.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
query_6_1 = f'''select 
                        count(v.id) vac_data_count
                from 
                        public.vacancies v
                where
                        v.name ilike '%data%' or v.name ilike '%данн%'
            '''

In [38]:
# результат запроса
pd.read_sql_query(query_6_1, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2603769340.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_1, connection)


vac_data_count
0            1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
query_6_2 = f'''select 
                        count(v.id) DS_Junior_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'   
                        and
                        (v.name ilike '%junior%' or 
                        v.experience='Нет опыта' or
                        v.employment='Стажировка')                     
            '''

In [40]:
# результат запроса
pd.read_sql_query(query_6_2, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2428688723.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_2, connection)


ds_junior_count
0               51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
query_6_3 = f'''select 
                        count(v.id) DS_SQL_postgres_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'   
                        and 
                        (v.key_skills ilike '%SQL%' or
                        v.key_skills ilike '%postgres%')                   
            '''

In [42]:
# результат запроса
pd.read_sql_query(query_6_3, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3902879435.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_3, connection)


ds_sql_postgres_count
0                    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
query_6_4 = f'''select 
                        count(v.id) DS_Python_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'   
                        and 
                        v.key_skills ilike '%Python%'                  
            '''

In [44]:
# результат запроса
pd.read_sql_query(query_6_4, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2548700122.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_4, connection)


ds_python_count
0              351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
query_6_5 = f'''select 
                        round(avg(length(v.key_skills)-length(replace(v.key_skills,chr(9),''))+1),2)
                        DS_key_skills_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'     
                        and
                        v.key_skills is not null              
            '''

In [46]:
# результат запроса
pd.read_sql_query(query_6_5, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3854041181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_5, connection)


ds_key_skills_count
0                 6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
query_6_6 = f'''select 
                        v.experience,
                        round(avg((coalesce(v.salary_from,v.salary_to) + coalesce(v.salary_to,v.salary_from))/2),0)
                        ds_avg_salary
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'      
                        and
                        (v.salary_from is not null or
                        v.salary_to is not null)  
                group by
                        v.experience        
            '''

In [48]:
# результат запроса
pd.read_sql_query(query_6_6, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1796641676.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_6, connection)


experience  ds_avg_salary
0           Нет опыта        74643.0
1  От 1 года до 3 лет       139675.0
2       От 3 до 6 лет       243115.0

***

#### выводы по предметному анализу
Основная масса вакансий в области DS предназначена для кандидатов с опытом. Крайне мало вакансий для начинающих.
SQL и postgres не являются преобладающими требованиями к кандидатам по сравнению с Python. 
Также стоит отметить, что достаточно насыщенным является перечень требований к ключевым навыкам (более 6 пунктов).
И стоит отметить, что для вакансий с опытом работы в DS от 3 до 6 лет указывается достаточно высокий 
средний уровень зарплаты, в районе 250т.р.

***

***

# Общий вывод по проекту

В базе представлено 49197 опубликованных вакансий от 23501 работодателя в 1362 регионах.
Вакансии относятся к одной или нескольким из 294 возможных сфер деятельности работодателя.
Почти для 20% не требуется никаких ключевых навыков, т.к. это требование заполнено только для 40741 вакансии.

Из всех вакансий ориентир по заработной плате указан менее чем для половины, только в 24073.
Полный диапазон зарплат (от и до) обозначен только в 12690 вакансиях.
Диапазон зарплат "от" и "до" по среднеарифметическим значениям выглядит так: от 71065 до 110537.

Распределение вакансий по регионам зависит от населенности и развитости региона, лидером является 
Москва с более чем 10% от всего количества вакансий (5333 вакансии в Москве), с разницей почти в 2 раза 
относительно следующего за ней Санкт-Петербурга (2851 вакансия в СПб).

В более чем 90% случаев требуется полная занятость. При этом, с полной занятостью нужен сотрудник на полный день 
в 35367 случаях, удаленная работа предлагается в 7802 вакансиях, и гибкий график допустим в 1593 вакансии.

Наиболее востребованными (более половины, 26152 вакансии) являются сотрудники с опытом работы от 1 года до 3 лет.
В 30% случаев (14511 вакансий) работодатели рассматривают кандидатов с опытом от 3 до 6 лет. 
Для кандидатов без опыта подходит около 15% вакансий (7197).

Явным лидером по количеству размещенных вакансий является Яндекс (почти 4% от общего количества), 
что объясняется возможностью удаленной работы и связанным с этим дублированием вакансий по множеству регионов.
У следующих за ним работодателей Ростелеком, Тинькофф и СБЕР размещено приблизительно в 4 раза меньше вакансий, 
чем у Яндекса.
Тот же Яндекс является лидером и по количеству регионов, в которых публикует вакансии (181 регион). 
Следующим идет тот же Ростелеком, что и по количеству вакансий в целом (152 региона).

Ниже представлены дополнительные исследования и вывод по той части вакансий, которая имеет отношение к IT.

***

In [49]:
# оценим общее количество вакансий DS без ограничения для грейда junior
query_total_1 = f'''select 
                        count(v.id) DS_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'                      
            '''
pd.read_sql_query(query_total_1, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2193350224.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_1, connection)


ds_count
0       480

In [50]:
# оценим среднее количество ключевых навыков на одну вакансию для всей базы (с заполненными требованиями в этом поле)
query_total_2 = f'''select 
                        round(avg(length(v.key_skills)-length(replace(v.key_skills,chr(9),''))+1),2)
                        key_skills_count
                from 
                        public.vacancies v
                where
                        v.key_skills is not null              
            '''
pd.read_sql_query(query_total_2, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/879936488.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_2, connection)


key_skills_count
0              6.37

In [51]:
# оценим лидеров-работодателей по количеству вакансий DS
query_total_3 = f'''select 
                        distinct e.name employer,
                        count(v.id) vac_count
                from 
                        public.vacancies v
                        left join public.employers e on v.employer_id=e.id
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'     
                group by
                        1
                order by
                        2 desc
                limit(10)
            '''
pd.read_sql_query(query_total_3, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/3862752397.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_3, connection)


employer  vac_count
0                   СБЕР         37
1        Bell Integrator         25
2         Банк ВТБ (ПАО)         18
3                     VK         15
4  Positive Technologies         11
5         EvenBet Gaming          9
6                 Яндекс          9
7                МегаФон          8
8               Andersen          7
9                   Ozon          6

In [52]:
# оценим лидеров-работодателей по количеству вакансий DS
query_total_4 = f'''select 
                        distinct e.name employer,
                        count(v.id) vac_count
                from 
                        public.vacancies v
                        left join public.employers e on v.employer_id=e.id
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'    
                        and
                        v.experience='Нет опыта' 
                group by
                        e.id
                order by
                        2 desc
            '''
pd.read_sql_query(query_total_4, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1741440280.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_4, connection)


employer  vac_count
0                                              Comexp          3
1                                       SteadyControl          2
2    The Skolkovo Institute of Science and Technology          2
3                                              Диплей          2
4                                                DINS          1
5                                   DNS Головной офис          1
6                                                Ozon          1
7                             Алгоритмические Системы          1
8                                          АЛЬФА-БАНК          1
9                                             Апэрбот          1
10                          БСТ Менеджмент-Консалтинг          1
11  КБ Уральский банк реконструкции и развития (УБ...          1
12                                            МегаФон          1
13                           Миррико, Группа Компаний          1
14                                           ННФормат          1
15                                               СБЕР          1
16      СтарЛайн, Научно-производственное объединение          1
17                                  Строительный Двор          1
18                                      Улыбка радуги          1
19                                   Университет ИТМО          1
20  Федеральное автономное учреждение Государствен...          1
21                                   Целус Менеджмент          1
22                                          Шлюмберже          1

In [53]:
# посмотрим, в каких регионах наибольшая представленность по вакансиям DS, а также оценим общее количество регионов
query_total_5 = f'''select 
                        a.name area,
                        count(v.area_id) cnt
                from 
                        public.vacancies v
                        left join public.areas a on v.area_id = a.id
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'
                group by 
                        a.name
                order by 
                        count(v.area_id) desc
                limit(10)
            '''
pd.read_sql_query(query_total_5, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/1362849849.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_5, connection)


area  cnt
0           Москва  217
1  Санкт-Петербург   64
2      Новосибирск   23
3  Нижний Новгород   20
4           Казань   16
5           Алматы   16
6            Минск    9
7            Томск    8
8     Екатеринбург    7
9          Армения    6

In [54]:
# оценим тип графика и вид трудоустройства для вакансий DS
query_total_6 = f'''select 
                        v.schedule,
                        v.employment,
                        count(v.id) vacancies_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'
                group by 
                        v.schedule,
                        v.employment
                order by 
                        count(v.id) desc
            '''
pd.read_sql_query(query_total_6, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/469996741.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_6, connection)


schedule           employment  vacancies_count
0       Полный день     Полная занятость              319
1  Удаленная работа     Полная занятость              110
2     Гибкий график     Полная занятость               37
3     Гибкий график  Частичная занятость                4
4  Удаленная работа  Частичная занятость                3
5       Полный день           Стажировка                2
6  Удаленная работа     Проектная работа                2
7     Гибкий график           Стажировка                2
8       Полный день  Частичная занятость                1

In [55]:
query_total_7 = f'''select 
                        distinct v.experience,
                        count(v.id) vac_count
                from 
                        public.vacancies v
                where
                        (v.name ilike '%data scientist%' or 
                        v.name ilike '%data science%' or
                        v.name ilike '%исследователь данных%' or
                        v.name like '%ML%' or
                        v.name ilike '%machine learning%' or
                        v.name ilike '%машинн%обучен%')
                        and
                        v.name not ilike '%html%'
                group by
                        1
                order by
                        2 desc
            '''
pd.read_sql_query(query_total_7, connection)

/var/folders/0v/d10yh6tj1mn387jzy55lmxv80000gn/T/ipykernel_10315/2367818742.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_total_7, connection)


experience  vac_count
0  От 1 года до 3 лет        223
1       От 3 до 6 лет        195
2         Более 6 лет         34
3           Нет опыта         28

***

Применительно к IT-индустрии можно отметить следующие моменты.
В сфере "Разработка ПО" заявили себя 15% работодателей (3553 из 23501).
Из них только около половины, судя по описанию вакансии, имеют отношение к данным (1771 вакансия).

480 вакансий DS разместили 225 работодателей, т.е. среднее количество вакансий чуть больше 2 на работодателя. 
Имеются ярко выраженный лидер по количеству вакансий - СБЕР (37 вакансий). К 10-му в рейтинге работодателю 
количество вакансий снижается уже до 6.

Вакансии DS представлены в 52 регионах, при этом 217 из 480 вакансий размещены в Москве, еще 64 в СПб, 
23 в Новосибирске, 20 в Нижнем Новгороде, в остальных регионах вакансий менее 20.

Рейтинг вакансий по графику работы и виду трудоустройства повторяет общую картину. 
В 97% случаев (466 из 480 вакансий) работодатели хотят видеть своих работников на условиях полной занятости. 
Из них 319 на полный день в офисе, 110 готовы предоставить удаленку и только 37 готовы на гибкий график.

Очень мало вариантов для начинающих DS, всего 51 вакансия (чуть больше 10% от 480 вакансий DS независимо от грейда),
если ориентироваться на описание вакансии.
В то же время, исходя из требований к опыту, всего 28 вакансий, доступных для соискателей без опыта, от 22 работодателей. 
Сопоставимое количество вакансий для опыта работы от 1 года до 3 лет (223 вакансии) и от 3 до 6 лет (195 вакансий).
Из всех вакансий DS для 201 (42%) требуются навыки SQL или postgres, а для 351 вакансии (73%) - навык работы в Python.

Вакансии DS не отличаются по среднему количеству затребованных ключевых навыков от остальных вакансий, 
в среднем в описании вакансии их около 6,4 и для DS и в целом для заполненных требований к ключевым навыкам по всей базе.

Средняя зарплата DS сильно зависит от опыта работы. От 75т.р. для начинающего, к 140т.р. с опытом от 1 года до 3 лет
и до 243т.р., если опыт составляет от 3 до 6 лет.

Резюмируя, хочется отметить, что исходя из вышеуказанной информации, вход в сферу DS сложный для новичков, 
поскольку количество соответствующих вакансий стремится к нулю. Основная масса предложений для имеющих опыт кандидатов 
приходится на Москву, полный день в офисе с полной занятостью. Возможно, отчасти с этим связаны финансовые условия
для кандидатов, т.к. средний уровень ЗП для Москвы в целом гораздо выше среднего по России.